src/System.Text.JsonLab/System/Text/Json/JsonReaderState.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.JsonLab
+{
+    public struct JsonReaderState : IEquatable<JsonReaderState>
+    {
+        // This mask represents a tiny stack to track the state during nested transitions.
+        // The first bit represents the state of the current level (1 == object, 0 == array).
+        // Each subsequent bit is the parent / containing type (object or array). Since this
+        // reader does a linear scan, we only need to keep a single path as we go through the data.
+        internal ulong _containerMask;
+        internal Stack<bool> _stack;
+        internal bool _inObject;
+        internal int _depth;
+        internal JsonTokenType _tokenType;
+        internal bool _searchedNextLast;
+
+        public override bool Equals(object obj)
+        {
+            return obj is JsonReaderState && Equals((JsonReaderState)obj);
+        }
+
+        public bool Equals(JsonReaderState other)
+        {
+            return _containerMask == other._containerMask &&
+                   EqualityComparer<Stack<bool>>.Default.Equals(_stack, other._stack) &&
+                   _inObject == other._inObject &&
+                   _depth == other._depth &&
+                   _tokenType == other._tokenType &&
+                   _searchedNextLast == other._searchedNextLast;
+        }
+
+        public override int GetHashCode()
```

Comment:
This implementation was generated automatically by VS. What's the common pattern for implementing GetHashCode?

Reply:
Going forward, we have APIs to create hashcodes, but these are not yet in NetStandard. 
This type will probably not be used in hash tables, and so I would just simplify what you have. For example, I would remove _stack and the bools (_inObject and _searchedNextLast). The stack is too complex and bools probably don't add much entropy to the hash.